File: ios/Classes/coins/NEAR.swift
/*
 NEAR
 */
import WalletCore

class NEAR: Coin  {
    init() {
        super.init(name: "NEAR", coinType: .near)
    }
    
    override func getPublicKey(path: String, mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        let publicKey: String? = wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeyEd25519().data.base64EncodedString()
        return publicKey
    }

    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        
        let input = NEARSigningInput.with {
                    $0.signerID = txData["signerID"] as! String
                    $0.nonce = txData["nonce"] as! UInt64
                    $0.receiverID = txData["receiverID"] as! String

                    $0.actions = [
                        NEARAction.with({
                            $0.transfer = NEARTransfer.with {
                                // uint128_t / little endian byte order
                                $0.deposit = Data(hexString: txData["amount"] as! String)!
                            }
                        }),
                    ]

                    $0.blockHash = Base58.decodeNoCheck(string: txData["blockHash"] as! String)!
                    $0.privateKey = privateKey!.data
                }

                let output: NEARSigningOutput = AnySigner.sign(input: input, coin: .near)
        return output.signedTransaction.hexString
    }
}

File: android/src/main/kotlin/africa/ejara/trustdart/coins/NEAR.kt
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.proto.NEAR
import wallet.core.jni.proto.NEAR.SigningOutput
import wallet.core.jni.Base58
import wallet.core.jni.HDWallet
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.base64String
import africa.ejara.trustdart.utils.toLong

class NEAR : Coin("NEAR", CoinType.NEAR) {

    override fun getPublicKey(path: String, mnemonic: String, passphrase: String): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path).publicKeyEd25519.data().base64String()
    }

    override fun getRawPublicKey(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path).publicKeyEd25519.data()
    }

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val secretKey = wallet.getKey(coinType, path)
        val transferAction = NEAR.Transfer.newBuilder().apply {
            deposit =
                ByteString.copyFrom(Numeric.hexStringToByteArray((txData["amount"] as String)))
        }.build()
        val signingInput = NEAR.SigningInput.newBuilder().apply {
            signerId = txData["signerID"] as String
            nonce = txData["nonce"]!!.toLong()
            receiverId = txData["receiverID"] as String
        //    addActionsBuilder().apply {
        //        transfer = transferAction
        //    }
            blockHash = ByteString.copyFrom(Base58.decodeNoCheck(txData["blockHash"] as String))
            privateKey = ByteString.copyFrom(secretKey.data())
        }.build()

        val output = AnySigner.sign(signingInput, CoinType.NEAR, SigningOutput.parser())
        return Numeric.toHexString(output.signedTransaction.toByteArray())
    }

}